packages/angular-query-experimental/src/util/create-injection-token/create-injection-token.ts

```diff
@@ -0,0 +1,268 @@
+ * @returns {CreateInjectionTokenReturn}
+ *
+ * @example
+ * ```ts
+ * const [injectCounter, provideCounter, COUNTER] = createInjectionToken(() => signal(0));
+ *
+ * export class Counter {
+ *  counter = injectCounter(); // WritableSignal<number>
+ * }
+ * ```
+ */
+export function createInjectionToken<
+  TFactory extends (...args: Array<any>) => any,
+  TFactoryDeps extends Parameters<TFactory> = Parameters<TFactory>,
+  TOptions extends CreateInjectionTokenOptions<
+    TFactory,
+    TFactoryDeps
+  > = CreateInjectionTokenOptions<TFactory, TFactoryDeps>,
+  TFactoryReturn = TOptions['multi'] extends true
+    ? Array<ReturnType<TFactory>>
+    : ReturnType<TFactory>,
+>(
+  factory: TFactory,
+  options?: TOptions,
+): CreateInjectionTokenReturn<TFactoryReturn> {
+  const tokenName = factory.name || factory.toString();
+  const opts =
+    options ??
+    ({ isRoot: true } as CreateInjectionTokenOptions<TFactory, TFactoryDeps>);
+
+  opts.isRoot ??= true;
+
+  // NOTE: multi tokens cannot be a root token. It has to be provided (provideFn needs to be invoked)
+  // for the 'multi' flag to work properly
+  if (opts.multi) {
+    opts.isRoot = false;
+  }
+
+  if (opts.isRoot) {
+    if (opts.token) {
+      throw new Error(`\
+createInjectionToken is creating a root InjectionToken but an external token is passed in.
+`);
+    }
+
+    const token = new InjectionToken<TFactoryReturn>(`Token for ${tokenName}`, {
+      factory: () => {
+        if (opts.deps && Array.isArray(opts.deps)) {
+          return factory(...opts.deps.map((dep) => inject(dep)));
+        }
+        return factory();
+      },
+    });
+
+    const injectFn = createInjectFn(
+      token,
+    ) as CreateInjectionTokenReturn<TFactoryReturn>[0];
+
+    return [
+      injectFn,
+      createProvideFn(
+        token,
+        factory,
+        opts as CreateProvideFnOptions<TFactory, TFactoryDeps>,
+      ) as CreateInjectionTokenReturn<TFactoryReturn>[1],
+      token,
+      () => ({
+        provide: ENVIRONMENT_INITIALIZER,
+        useValue: () => injectFn(),
+        multi: true,
+      }),
+    ];
+  }
+
+  const token =
+    opts.token || new InjectionToken<TFactoryReturn>(`Token for ${tokenName}`);
+  return [
+    createInjectFn(token) as CreateInjectionTokenReturn<TFactoryReturn>[0],
+    createProvideFn(
+      token,
+      factory,
+      opts as CreateProvideFnOptions<TFactory, TFactoryDeps>,
+    ) as CreateInjectionTokenReturn<TFactoryReturn>[1],
+    token,
+    () => [],
+  ];
+}
+
+export function createNoopInjectionToken<
+  TValue,
+  TMulti extends boolean = false,
+  TOptions = Pick<
+    CreateInjectionTokenOptions<() => void, []>,
+    'extraProviders'
+  > &
+    (TMulti extends true ? { multi: true } : Record<string, never>),
+>(description: string, options?: TOptions) {
+  type TReturn = TMulti extends true ? Array<TValue> : TValue;
+
+  const token =
+    (options as CreateInjectionTokenOptions<() => void, []>).token ||
```

Comment:
It looks like the test failure is related to this line, `options` is an optional parameter to `createNoopInjectionToken`, so it just needs an optional chaining operator:

```suggestion
    (options as CreateInjectionTokenOptions<() => void, []>)?.token ||
```